gateway/crates/gateway-binary/tests/telemetry/metrics/operation/subgraph.rs

```diff
@@ -0,0 +1,242 @@
+            {
+              "message": "Request to subgraph 'accounts' failed with: error sending request for url (http://127.0.0.1:46697/)",
+              "path": [
+                "me"
+              ],
+              "extensions": {
+                "code": "SUBGRAPH_REQUEST_ERROR"
+              }
+            }
+          ]
+        }
+        "###);
+
+        tokio::time::sleep(METRICS_DELAY).await;
+
+        let row = clickhouse
+            .query(
+                r#"
+                SELECT Count, Attributes
+                FROM otel_metrics_exponential_histogram
+                WHERE ServiceName = ? AND StartTimeUnix >= ?
+                    AND ScopeName = 'grafbase'
+                    AND MetricName = 'graphql.subgraph.request.body.size'
+                "#,
+            )
+            .bind(&service_name)
+            .bind(start_time_unix)
+            .fetch_one::<ExponentialHistogramRow>()
+            .await
+            .unwrap();
+
+        insta::assert_json_snapshot!(row, @r###"
+        {
+          "Count": 1,
+          "Attributes": {
+            "graphql.subgraph.name": "accounts"
+          }
+        }
+        "###);
+    });
+}
+
+#[test]
+fn retries() {
+    let config = indoc::indoc! {r#"
+        [gateway.retry]            
+        enabled = true
+        min_per_second = 1
+        ttl = "1s"
+        retry_percent = 0.1
+        retry_mutations = false
+    "#};
+
+    with_custom_gateway(
+        config,
+        |service_name, start_time_unix, gateway, clickhouse| async move {
+            let response = gateway
+                .gql::<serde_json::Value>("query Simple { me { id } }")
+                .send()
+                .await;
+
+            insta::assert_json_snapshot!(response, @r###"
+            {
+              "data": null,
+              "errors": [
+                {
+                  "message": "Request to subgraph 'accounts' failed with: error sending request for url (http://127.0.0.1:46697/)",
+                  "path": [
+                    "me"
+                  ],
+                  "extensions": {
+                    "code": "SUBGRAPH_REQUEST_ERROR"
+                  }
+                }
+              ]
+            }
+            "###);
+
+            tokio::time::sleep(METRICS_DELAY).await;
+
+            let rows = clickhouse
+                .query(
+                    r#"
+                SELECT Value, Attributes
+                FROM otel_metrics_sum
+                WHERE ServiceName = ? AND StartTimeUnix >= ?
+                    AND ScopeName = 'grafbase'
+                    AND MetricName = 'graphql.subgraph.request.retries'
+                "#,
+                )
+                .bind(&service_name)
+                .bind(start_time_unix)
+                .fetch_all::<SumRow>()
+                .await
+                .unwrap();
+
+            insta::assert_json_snapshot!(rows, @r###"
+            [
+              {
+                "Value": 1.0,
```

Comment:
Weird, do we expect to retry only once with that config? I guess so, since it's a minimum of 1 per second, and we haven't put any successful request in the budget yet at this point. Kind of makes sense, but I am still having trouble wrapping my head around how the budget works at the very beginning sometimes.

Reply:
So what happens here is that it retries once (counter +1), then sleeps, then tries to retry the second time but has no balance in the bucket and aborts. Two rows to the database.

Reply:
The retry counter has that aborted flag, which you can either filter out or have as a separate message, which then tells you that this retry stopped and will not continue.